oracle_fdw.c

```diff
@@ -821,9 +822,63 @@ oracleGetForeignRelSize(PlannerInfo *root, RelOptInfo *baserel, Oid foreigntable
 				fdwState->oraTable->cols[i]->used = 1;
 	}
 
-	/* construct Oracle query and get the list of parameters and actions for RestrictInfos */
-	fdwState->query = createQuery(fdwState, baserel, for_update, root->query_pathkeys);
-	elog(DEBUG1, "oracle_fdw: remote query is: %s", fdwState->query);
+	/* createQuery is moved to GetForeignPlan. */
+	/* fdwState->for_update is needed by createQuery in GetForeignPlans */
+	fdwState->for_update = for_update;
+	/* fdwState->usable_pathkeys is needed by create_foreign_scan in GetForeignPaths. */
+	/* So it is created in this place. */
+
+	/*
+	 * Determine whether we can potentially push query pathkeys to the remote
+	 * side, avoiding a local sort.
+	 */
+	ListCell *cell;
+	foreach(cell, root->query_pathkeys)
+	{
+		PathKey *pathkey = (PathKey *)lfirst(cell);
+		EquivalenceClass *pathkey_ec = pathkey->pk_eclass;
+		Expr *em_expr = NULL;
+		char *sort_clause;
+		Oid em_type;
+		bool can_pushdown;
+
+		/*
+		 * deparseExpr would detect volatile expressions as well, but
+		 * ec_has_volatile saves some cycles.
+		 */
+		can_pushdown = !pathkey_ec->ec_has_volatile
+				&& ((em_expr = find_em_expr_for_rel(pathkey_ec, baserel)) != NULL);
+
+		if (can_pushdown)
+		{
+			em_type = exprType((Node *)em_expr);
+
+			/* expressions of a type different from this are not safe to push down into ORDER BY clauses */
+			if (em_type != INT8OID && em_type != INT2OID && em_type != INT4OID && em_type != OIDOID
+					&& em_type != FLOAT4OID && em_type != FLOAT8OID && em_type != NUMERICOID && em_type != DATEOID
+					&& em_type != TIMESTAMPOID && em_type != TIMESTAMPTZOID && em_type != INTERVALOID)
+				can_pushdown = false;
+		}
+
+		if (can_pushdown &&
+			((sort_clause = deparseExpr(fdwState->session, baserel, em_expr, fdwState->oraTable, &(fdwState->params))) != NULL)){
```

Comment:
The local variable `sort_clause` is not used at all.
See my comment at line 2547.